magma/circuit.py

```diff
@@ -921,3 +923,32 @@ def __new__(metacls, name, bases, dct):
 
 class DebugCircuit(Circuit, metaclass=DebugDefineCircuitKind):
     pass
+
+
+InstanceCallback = Callable[[List[CircuitType], Dict], None]
+
+
+def register_instance_callback(
+        instance: CircuitType,
+        callback: InstanceCallback,
+):
+    try:
+        get_instance_callback(instance)
+    except AttributeError:
+        pass
+    else:
+        raise AttributeError(
+            f"Instance {instance} already has callback registered"
+        )
```

Comment:
```suggestion
    if hasattr(instance, "_callback_"):
        raise AttributeError(
            f"Instance {instance} already has callback registered"
        )
```

Perhaps a `hasattr` would be better here? Less code and I don't think going with an eager try/catch provides any performance benefit since we still have to check the attribute anyways.

Reply:
:+1: